src/score_gen/score.rs

```diff
@@ -1,162 +1,542 @@
+pub struct Score {
+    pub(crate) score: Scalar,
+    pub(crate) y: Scalar,
+    pub(crate) y_prime: Scalar,
+    pub(crate) r1: Scalar,
+    pub(crate) r2: Scalar,
 }
 
-/// Computes the score of the bid printing in the ConstraintSystem the proof of the correct
-/// obtention of the score.
-///
-/// Takes 3615 constraints.
-pub fn compute_score_gadget(
-    composer: &mut StandardComposer,
-    bid: &Bid,
-    bid_value: Variable,
-    secret_k: Variable,
-    bid_tree_root: Variable,
-    consensus_round_seed: Variable,
-    latest_consensus_round: Variable,
-    latest_consensus_step: Variable,
-) -> Variable {
+impl Score {
+    pub(crate) fn new(score: Scalar, y: Scalar, y_prime: Scalar, r1: Scalar, r2: Scalar) -> Self {
+        Score {
+            score,
+            y,
+            y_prime,
+            r1,
+            r2,
+        }
+    }
+}
+
+/// Given a `Bid`, compute it's Score and return it.
+pub(crate) fn compute_score(bid: &Bid) -> Result<Score, Error> {
     // Compute `y` where `y = H(secret_k, Merkle_root, consensus_round_seed, latest_consensus_round, latest_consensus_step)`.
-    // This is done in ZK using the sponge hash gadget.
-    let y = sponge::sponge_hash_gadget(
-        composer,
-        &[
-            secret_k,
-            bid_tree_root,
-            consensus_round_seed,
-            latest_consensus_round,
-            latest_consensus_step,
-        ],
-    );
-    // We also need to compute the sponge hash with the scalars since we need to prove the correctness of the inverse
-    // of y'.
-    let scalar_y = sponge::sponge_hash(&[
-        // We wrap up the JubJubScalar as a BlsScalar which will always fit.
-        // That means that the unwrap is safe.
-        Scalar::from_bytes(&bid.secret_k.to_bytes()).unwrap(),
+    let y = sponge::sponge_hash(&[
+        bid.secret_k,
         bid.bid_tree_root,
         bid.consensus_round_seed,
         bid.latest_consensus_round,
         bid.latest_consensus_step,
     ]);
-    // Compute 1/y' where `y' = y & 2^129 -1`. This needs to be done for `Scalar` and `Variable` backends
-    // to then assert that the inverse is correct.
-    // Compute y' and 1/y'.
-    let (_, inv_y_prime_scalar) = compute_y_primes(scalar_y);
-
-    // Compute y' as a Variable.
-    let y_prime_var = {
-        let truncate_val =
-            composer.add_input(Scalar::from(2u64).pow(&[129, 0, 0, 0]) - Scalar::one());
-        composer.logic_and_gate(y, truncate_val, 256)
+
+    // Truncate Y to left 128 bits and interpret the result as 128-bit integer.
+    // Keep the right 128 bits as another integer (r1).
+    let r1 = BigUint::from_bytes_le(&y.to_bytes()[16..32]);
+    let y_prime = BigUint::from_bytes_le(&y.to_bytes()[0..16]);
+
+    // Get the bid value outside of the modular field and treat it as
+    // an integer.
+    let bid_value = BigUint::from_bytes_le(&bid.value.to_bytes());
+    // Compute the final score
+    let (f, r2) = match y_prime == BigUint::zero() {
```

Comment:
Will be addessed in #15 's PR

Reply:
For the future, notice that this kind of stuff should be taken care in the same PR since they're not a change of logic or refactoring: it's part of the review process. In this case for example it was not only verbose and unnecessary using `match` but it also was inconsistent since later in the code you actually use the `if` with the same purpose.

We should never merge code with this kind of inconsistency.

I didn't ask for changes because overall looks good and there was no need to address all of the nits, but at least this kind of things should be taken care of before merging.

Reply:
I've been looking at it . And looks clearer to me in this way TBH. It's more clear what happens on each case.

Reply:
Again, just for future reference; one problem is the inconsistency as mentioned if have both `match` and `if` in the same code, so we should never have this kind of inconsistency for the same operation: stick to one.

Second of all, `if` it's exactly made for this kind of scenario where you have just to check boolean value. It's really weird using `match` to check boolean value. You can do that, but it's weird. At this point you should never use an `if` statement if `match` looks "clearer" to you, but always a `match`. :)